pkg/sshd/server.go

```diff
@@ -194,85 +233,85 @@ func (s *server) answer(channel ssh.Channel, requests <-chan *ssh.Request, sshCo
 			clean := cleanExec(req.Payload)
 			parts := strings.SplitN(clean, " ", 2)
-
-			router := s.c.Get("cookoo.Router", nil).(*cookoo.Router)
-
-			// TODO: Should we unset the context value 'cookoo.Router'?
-			// We need a shallow copy of the context to avoid race conditions.
-			cxt := s.c.Copy()
-			cxt.Put("SSH_CONNECTION", sshConn)
-
-			// Only allow commands that we know about.
 			switch parts[0] {
 			case "ping":
-				cxt.Put("channel", channel)
-				cxt.Put("request", req)
-				sshPing := cxt.Get("route.sshd.sshPing", "sshPing").(string)
-				err := router.HandleRequest(sshPing, cxt, true)
+				err := Ping(channel, req)
 				if err != nil {
-					log.Warnf(s.c, "Error pinging: %s", err)
+					log.Info("Error pinging: %s", err)
 				}
 				return err
 			case "git-receive-pack", "git-upload-pack":
 				if len(parts) < 2 {
-					log.Warn(s.c, "Expected two-part command.\n")
+					log.Info("Expected two-part command.")
 					req.Reply(ok, nil)
 					break
 				}
-
-				repoName := parts[1]
-				if err := s.pushLock.Lock(repoName, time.Duration(0)); err != nil {
-					log.Errf(s.c, multiplePush)
+				repoName, err := cleanRepoName(parts[1])
+				if err != nil {
+					log.Err("Illegal repo name: %s.", err)
+					channel.Stderr().Write([]byte("No repo given"))
+					return err
+				}
+				if err = s.pushLock.Lock(repoName, time.Duration(0)); err != nil {
+					log.Info(multiplePush)
 					// The error must be in git format
 					if err := gitPktLine(channel, fmt.Sprintf("ERR %v\n", multiplePush)); err != nil {
-						log.Errf(s.c, "Failed to write to channel: %s", err)
+						log.Err("Failed to write to channel: %s", err)
 					}
 					sendExitStatus(1, channel)
 					req.Reply(false, nil)
 					return nil
 				}
 
 				req.Reply(true, nil) // We processed. Yay.
-
-				cxt.Put("channel", channel)
-				cxt.Put("request", req)
-				cxt.Put("operation", parts[0])
-				cxt.Put("repository", parts[1])
-				sshGitReceive := cxt.Get("route.sshd.sshGitReceive", "sshGitReceive").(string)
-				err := router.HandleRequest(sshGitReceive, cxt, true)
-				if err := s.pushLock.Unlock(repoName, time.Duration(0)); err != nil {
-					log.Errf(s.c, "unable to unlock repository lock for %s (%s)", repoName, err)
-					// TODO: this is an important error case that needs to be covered
-					// Probably the best solution is to change the lock into a lease so that even on unlock
-					// failures, RepositoryLock will eventually yield
+				if !strings.Contains(sshconn.Permissions.Extensions["apps"], repoName) {
+					if err := s.pushLock.Unlock(repoName, time.Duration(0)); err != nil {
+						log.Err("unable to unlock repository lock for %s (%s)", repoName, err)
+						// TODO: this is an important error case that needs to be covered
+						// Probably the best solution is to change the lock into a lease so that even on unlock
+						// failures, RepositoryLock will eventually yield
+					}
+					return fmt.Errorf("user has no permission to build the app")
```

Comment:
this error should be declared globally using `errors.New("user has no permission to build the app")`